testdata/schema.go

```diff
@@ -0,0 +1,232 @@
+			newFields = append(newFields[:i], newFields[i+1:]...)
+			i--
+		}
+	}
+	return newFields
+}
+
+func removeFieldsByType(fields []arrow.Field, t ...arrow.Type) []arrow.Field {
+	newFields := make([]arrow.Field, len(fields))
+	copy(newFields, fields)
+	for _, d := range t {
+		for i := 0; i < len(newFields); i++ {
+			if newFields[i].Type.ID() == d {
+				newFields = append(newFields[:i], newFields[i+1:]...)
+				i--
+			}
+		}
+	}
+	return newFields
+}
+
+func removeFieldsByDataType(fields []arrow.Field, dt ...arrow.DataType) []arrow.Field {
+	newFields := make([]arrow.Field, len(fields))
+	copy(newFields, fields)
+	for _, d := range dt {
+		for i := 0; i < len(newFields); i++ {
+			if arrow.TypeEqual(newFields[i].Type, d) {
+				newFields = append(newFields[:i], newFields[i+1:]...)
+				i--
+			}
+		}
+	}
+	return newFields
+}
+
+// ListOfFields returns a list of fields that are lists of the given fields.
+func ListOfFields(baseFields []arrow.Field) []arrow.Field {
+	fields := make([]arrow.Field, len(baseFields))
+	for i := 0; i < len(baseFields); i++ {
+		fields[i] = arrow.Field{Name: baseFields[i].Name + "_list", Type: arrow.ListOf(baseFields[i].Type), Nullable: true}
+	}
+	return fields
+}
+
+// MapOfFields returns a list of fields that are maps of the given fields.
+func MapOfFields(baseFields []arrow.Field) []arrow.Field {
+	fields := make([]arrow.Field, len(baseFields))
+	for i := 0; i < len(baseFields); i++ {
+		fields[i] = arrow.Field{Name: baseFields[i].Name + "_map", Type: arrow.MapOf(baseFields[i].Type, baseFields[i].Type), Nullable: true}
+	}
+	return fields
+}
+
+// TestSourceOptions controls which types are included by TestSourceFields.
+type TestSourceOptions struct {
+	IncludeLists      bool // lists of all primitive types. Lists that were supported by CQTypes are always included.
+	IncludeTimestamps bool // timestamp types. Microsecond timestamp is always be included, regardless of this setting.
+	IncludeDates      bool
+	IncludeMaps       bool
+	IncludeStructs    bool
+	IncludeIntervals  bool
+	IncludeDurations  bool
+	IncludeTimes      bool // time of day types
+	IncludeLargeTypes bool // e.g. large binary, large string
+}
+
+// TestSourceFields returns fields for all Arrow types and composites thereof. TestSourceOptions controls
+// which types are included.
+func TestSourceFields(opts TestSourceOptions) []arrow.Field {
+	// cq fields
+	var cqFields []arrow.Field
+	cqIDMetadata := arrow.NewMetadata([]string{schema.MetadataUnique}, []string{"true"})
+	cqFields = append(cqFields, arrow.Field{Name: schema.CqIDColumn.Name, Type: types.NewUUIDType(), Nullable: false, Metadata: cqIDMetadata})
+	cqFields = append(cqFields, arrow.Field{Name: schema.CqParentIDColumn.Name, Type: types.NewUUIDType(), Nullable: false})
+
+	var basicFields []arrow.Field
+	basicFields = append(basicFields, PrimitiveFields()...)
+	basicFields = append(basicFields, BinaryFields()...)
+	basicFields = append(basicFields, FixedWidthFields()...)
```

Comment:
Need to add both nullable and not nullable fields of the same type to the schema

Reply:
This comment is the same as the one above?

Reply:
yes